vim25/soap/json_client.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright (c) 2023-2023 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package soap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"reflect"
+	"strings"
+
+	"github.com/vmware/govmomi/vim25/xml"
+
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+const (
+	sessionHeader = "vmware-api-session-id"
+)
+
+var (
+	// ErrInvalidResponse is used during unmarshaling when the response content
+	// does not match expectations e.g. unexpected HTTP status code or MIME
+	// type.
+	ErrInvalidResponse error = errors.New("Invalid response")
```

Comment:
It would be nice to have standard set of errors that we can use with errors.Is, errors.As. This was kind of broken before 1.20 https://tip.golang.org/doc/go1.20#errors